### ISSUE TITLE
Robot Framework and LLMs for Enhanced Testing

### ISSUE THREAD

**Issue by raballew** [jumpstarter-dev/jumpstarter#73]
- Testing for embedded systems often involves complex test scripts written directly in programming languages like Python, requiring specialized expertise.
- This complexity can limit the involvement of non-programmers (e.g., system engineers, manual testers, domain experts) in reviewing, understanding, or contributing to test case development.
- Maintaining large suites of code-based HIL tests can be time-consuming and challenging.
- There's a growing interest in leveraging Large Language Models (LLMs) to assist with test generation, analysis, and debugging, but applying them directly to complex HIL test code can be difficult.
- Managing the underlying HIL infrastructure (device access, control, connectivity) alongside test logic adds another layer of complexity.

**Proposal**

- Integrate Robot Framework (RF) as the primary (?!?) test authoring and execution layer on top of Jumpstarter (JS).
    - Utilize RF's keyword-driven, plain-English syntax for defining HIL test cases.
    - Develop a custom RF test library (in Python) that acts as a "glue" layer, translating RF keywords into API calls to the Jumpstarter system.
- Utilize JS for its core HIL capabilities:
    - Hardware abstraction and device control (power, flashing, console access) via JS drivers.
    - Management of device connectivity, including potentially secure tunneling (e.g., for ADB, SSH access to the DUT).
    - Integration with underlying infrastructure (local machines, distributed Kubernetes setups).
- Integrate LLMs for AI-assisted testing:
    - Use InstructLab (or similar techniques) to fine-tune an LLM specifically for understanding and generating Robot Framework HIL test cases, leveraging RF's simpler syntax as training data.
    - Deploy the fine-tuned LLM using Ramalama (or similar techniques) for secure, containerized, and easily accessible model serving via a REST API.
    - Integrate the served LLM into developer workflows or CI/CD pipelines for tasks like:
        - Generating RF test steps based on natural language requirements.
        - Analyzing RF test logs for failure patterns or summarization.
        - Providing debugging assistance for RF test failures.

**Benefits**

- **Improved Test Readability & Maintainability**: RF's syntax makes tests easier to understand and maintain for a wider audience.
- **Increased Accessibility**: Lowers the barrier for non-programmers to participate in the testing process (review, potentially contribute).
- **Leverage RF Ecosystem**: Enables the use of existing RF libraries (Selenium, Appium, SSH, Database, etc.) within the JS HIL environment for complex end-to-end testing.
- **Enhanced Reporting**: RF provides detailed, user-friendly HTML logs and reports out-of-the-box.
- **Streamlined LLM Integration**: RF's structure is potentially more amenable to LLM training (InstructLab) and generation compared to raw Python code. Ramalama offers standardized, secure deployment.
- **Clear Separation of Concerns**: RF handles test logic/orchestration, while JS manages the HIL infrastructure and hardware interaction.

**Initial Thoughts**

- The custom RF "glue" library is critical and requires careful design regarding keyword granularity and error handling.
- The RF Remote Library architecture might be suitable for interacting with distributed Jumpstarter setups.
- Performance implications of the RF Remote Library's XML-RPC communication need careful evaluation, especially for latency-sensitive tests.
- Jumpstarter's tunneling capabilities should be leveraged to provide access (e.g., ADB, SSH) to the DUT for standard RF libraries like AppiumLibrary or SSHLibrary, configured to connect via the tunnel endpoint (likely localhost on the RF Remote Server host).
- LLM training would involve curating RF test examples, defining skills/taxonomy in InstructLab, generating synthetic data, and fine-tuning.
- Ramalama can simplify serving the tuned model, providing a consistent API endpoint for integration.

**Challenges**

- **Increased System Complexity**: Managing multiple frameworks (RF, JS, potentially InstructLab/Ramalama) and the interactions between them.
- **Glue Code Development & Maintenance**: Significant effort required to create and maintain the custom RF-JS library.
- **Performance Overhead**: Potential latency introduced by the RF Remote Library interface.
- **Debugging Complexity**: Failures could occur in multiple layers (RF test, RF core, glue library, RF Remote, network, JS components, hardware), making troubleshooting harder.
- **Maturity & Support**: This specific integration pattern is novel and lacks established community support or best practices.
- **Skill Requirements**: Teams need proficiency in RF, Python, Jumpstarter, and potentially LLM tooling.
- **LLM Effectiveness**: The actual benefit and accuracy of LLMs for generating complex HIL tests in RF remain to be proven.

**Comment by kirkbrauer** [comment]
## Proposal

Provide a standard Python decorator API for declaring the interface of a driver client that is abstracted to support multiple tools such as a CLI or Robot Framework keywords. This will enable driver authors to easily create a way for users to interact with their Python API without having to define specific interfaces for every possible tool we wish to integrate with.

## Key Benefits

- **Single Source of Truth:** Define driver functionalities once using common decorators that support most common use cases.
- **Multi-Interface Exposure:** Automatically generate CLI commands, Robot Framework keywords, and other interfaces from the decorated functions.
- **Extensibility:** Easily integrate with additional tools in the future without having to modify all driver clients to support them.
- **Decoupling:** Driver clients no longer need to take on a Click dependency to provide a CLI.

## Decorators

Decorators can be provided as part of the `jumpstarter` package as a `jumpstarter.client.deco` module.

### `@command`
Marks a Python function in the driver client class as a command.
```python3
@command("user-friendly-name", keyword="Custom Keyword", help="This command does something.")
def my_client_function(self):
    pass
```

Click CLI:
```shell
$ j user-friendly-name
```

Robot Framework:
```robot
*** Test Cases ***
My Test Case
    Custom Keyword
```

### `@option`
Specifies an option that is available to the user.
```python3
@cmd("user-friendly-name", keyword="Custom Keyword", help="This command does something.")
@option("my_param", "-p", "--param", help="This is a custom parameter.")
def my_client_function(self, my_param: str):
    pass
```

Click CLI:
```shell
$ j user-friendly-name --param something
```

Robot Framework:
```robot
*** Test Cases ***
My Test Case
    Custom Keyword    my_pram=something
```

### `@argument`
Specifies a positional argument that is available to the user
```python3
@cmd("user-friendly-name", keyword="Custom Keyword", help="This command does something.")
@argument("my_arg", help="This is a custom argument.")
def my_client_function(self, my_arg: str):
    pass
```

Click CLI:
```shell
$ j user-friendly-name ./some/file.txt
```

Robot Framework:
```robot
*** Test Cases ***
My Test Case
    Custom Keyword    ./some/file.txt
```

## Keyword/Command Discovery

To automatically provide an interface from the decorators, we will need to traverse the driver client tree to identify the available commands/keywords at runtime.

### Click CLI

We can use the same method we currently use to discover Click commands, however, instead of using the special `cli()`, method, we must instead inspect our custom decorators on public methods declared on the driver class. This will allow for re-use between the Python and CLI interfaces if needed to reduce the amount of code.

Nested drivers will create command groups in the same way that they currently do.

### Robot Framework

To discover Robot Framework keywords, we can also inspect the driver client class and generate "dynamic keywords" from it. This needs to be done at runtime and within the editor to provide Intellisense. (Separate proposal for language server integration). Jumpstarter can take care of all of the lifecycle and lease management as part of the dynamic driver library.

Nested driver clients will automatically prefix keywords with the driver alias. For example, the method `client.power.on()` would be converted to the keyword "Power On" where "On" is the keyword defined explicitly or implicitly by the decorator.

## Limitations

We cannot support nested CLIs due to the flat nature of the Jumpstarter driver client hierarchy. Additionally, more complex Click options that are not supported in frameworks such as Robot also cannot be used. If those options are needed, the user can use the raw click decorators instead to provide that functionality if needed.

### Interactive vs. Non-Interactive Sessions

We can introduce a `DriverClient` attribute called `self._is_interactive` to determine if the user is using an interactive shell or not. This can help when we need to prompt for user input but don't know if the driver client is running under Robot Framework or an interactive/non-interactive shell session.

### Logging

We can expose the Jumpstarter logging interface and pipe that output to the Robot Framework logger so those logs are correctly reflected and the log level is respected. A custom `Error` type can also be exposed to be automatically consumed by either Click or Robot Framework to correctly throw exceptions from within driver client methods.
